lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
+public:
+  /// Construct a new namesapce from a circuit op.  This namespace will be
+  /// composed of any operation in the first level of the circuit that contains
+  /// a symbol.
+  CircuitNamespace(Operation *op) {
+    op->walk([&](Operation *op) {
+      if (auto symbol =
+              op->getAttrOfType<StringAttr>(SymbolTable::getSymbolAttrName()))
+        internal.insert(symbol.getValue());
+      return WalkResult::skip();
+    });
   }
-  return Optional<AugmentedBundleType>(bundle);
-}
-
-/// Remove Grand Central Annotations associated with SystemVerilog interfaces
-/// that should emitted.  This pass works in three major phases:
-///
-/// 1. The circuit's annotations are examnined to figure out _what_ interfaces
-///    there are.  This includes information about the name of the interface
-///    ("defName") and each of the elements (sv::InterfaceSignalOp) that make up
-///    the interface.  However, no information about the _type_ of the elements
-///    is known.
-///
-/// 2. With this, information, walk through the circuit to find scattered
-///    information about the types of the interface elements.  Annotations are
-///    scattered during FIRRTL parsing to attach all the annotations associated
-///    with elements on the right components.
-///
-/// 3. Add interface ops and populate the elements.
-///
-/// Grand Central supports three "normal" element types and four "weird" element
-/// types.  The normal ones are ground types (SystemVerilog logic), vector types
-/// (SystemVerilog unpacked arrays), and nested interface types (another
-/// SystemVerilog interface).  The Chisel API provides "weird" elements that
-/// include: Boolean, Integer, String, and Double.  The SFC implementation
-/// currently drops these, but this pass emits them as commented out strings.
-struct GrandCentralPass : public GrandCentralBase<GrandCentralPass> {
-  void runOnOperation() override;
-
-  // A map storing mutable information about an element in an interface.  This
-  // is keyed using a (defName, name) tuple where defname is the name of the
-  // interface and name is the name of the element.
-  typedef DenseMap<std::pair<StringRef, StringRef>, ElementInfo> InterfaceMap;
 
 private:
-  // Store a mapping of interface name to InterfaceOp.
-  llvm::StringMap<sv::InterfaceOp> interfaces;
+  SmallString<0> newNameImpl(StringRef name) {
+    // Special case the situation where there is no name collision to avoid
+    // messing with the SmallString allocation below.
+    if (internal.insert(name).second)
+      return name;
+    size_t i = 0;
+    SmallString<0> tryName;
+    do {
+      tryName = (name + "_" + Twine(i++)).str();
+    } while (!internal.insert(tryName).second);
+    return tryName;
+  }
 
-  // Discovered interfaces that need to be constructed.
-  InterfaceMap interfaceMap;
+public:
+  /// Generate a new name and add it to the namespace.
+  SmallString<0> newName(StringRef name) { return newNameImpl(name); }
```

Comment:
- `std::string` > `SmallString<0>` everywhere - I'm not aware of a good reason to use SmallString<0>, am I missing something? Since `Twine.str()` is creating a `std::string` anyways, this should save a string copy.

- I would add a comment describing what it does, e.g. adds a number until it generates an unused name.

- You can delete `newNameImpl`and move body here, and the other caller becomes:
```diff
- SmallString<0> newName(Twine name) { return newNameImpl(name.str()); }
+ SmallString<0> newName(Twine name) { return newName(name.str()); }
```

Reply:
I think I was getting confused where `SmallVector<0>` is suggested as an alternative to `std::vector`. However, I don't see similar guidelines or usages of `SmallVector<0>`. I'll change this...